docs/application-tutorial/game_of_life.py

```diff
@@ -0,0 +1,551 @@
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_3b_b.reshape(1, 1, *weights_method_3b_b.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    grid_b = grid_a + convoluted_grid
+
+    # Apply a TLU: input in [0, 5], output in [0, 4]
+    grid_c = table_cp_next_cell_3b_b[grid_b]
+
+    # Add center
+    grid = grid_c + grid
+
+    # And a last TLU: input in [0, 5] and output in [0, 1]
+    grid = table_cp_next_cell_3b_c[grid]
+
+    return grid
+
+
+@fhe.compiler({"grid": "encrypted"})
+def update_grid_method_4b(grid):
+
+    # Method which uses a first TLU of 4 bits and a second TLU of 2 bits
+
+    # This is to workaround the fact that we have no pad option in fhe.conv
+    do_padded_fix = True
+
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_4b.reshape(1, 1, *weights_method_4b.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    grid_a = table_cp_next_cell_4b_a[convoluted_grid]
+    grid = grid_a + grid
+    grid = table_cp_next_cell_4b_b[grid]
+
+    return grid
+
+
+@fhe.compiler({"grid": "encrypted"})
+def update_grid_method_5b(grid):
+
+    # Method which uses a single TLU of 5 bits
+
+    # This is to workaround the fact that we have no pad option in fhe.conv
+    do_padded_fix = True
+
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_5b.reshape(1, 1, *weights_method_5b.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    grid = table_cp_next_cell_5b[convoluted_grid]
+
+    return grid
+
+
+@fhe.compiler({"grid": "encrypted"})
+def update_grid_basic(grid):
+
+    # Method which follows the naive approach
+
+    # This is to workaround the fact that we have no pad option in fhe.conv
+    do_padded_fix = True
+
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_basic.reshape(1, 1, *weights_method_basic.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    grid = table_cp_next_cell_basic_a[convoluted_grid] | (
+        table_cp_next_cell_basic_b[convoluted_grid] & (grid == 1)
+    )
+
+    return grid
+
+
+# Function for Game of Life
+def update_grid(grid, method="method_3b"):
+
+    assert grid.ndim == 4
+
+    if method == "method_basic":
```

Comment:
return instead of res and raise ValueError("Bad method") at the end

Reply:
ok done